src/secondary.rs

```diff
@@ -759,6 +796,114 @@ impl<'a, K: Key, V: 'a + Copy> Extend<(K, &'a V)> for SecondaryMap<K, V> {
     }
 }
 
+/// An Entry API
+#[derive(Debug)]
+pub enum Entry<'a, K: Key, V> {
+    /// The Occupied variant
+    Occupied(OccupiedEntry<'a, K, V>),
+    /// The Vacant variant
+    Vacant(VacantEntry<'a, K, V>),
+}
+
+impl<'a, K: Key, V> Entry<'a, K, V> {
+    /// a
+    pub fn or_insert(self, default: V) -> &'a mut V {
+        match self {
+            Entry::Occupied(x) => x.into_mut(),
+            Entry::Vacant(x) => x.insert(default),
+        }
+    }
+
+    /// a
+    pub fn or_insert_with<F: FnOnce() -> V>(self, default: F) -> &'a mut V {
+        match self {
+            Entry::Occupied(x) => x.into_mut(),
+            Entry::Vacant(x) => x.insert(default()),
+        }
+    }
+
+    /// a
+    pub fn and_modify<F>(self, f: F) -> Self
+        where
+            F: for<'b> FnOnce(&'b mut V) {
+        match self {
+            Entry::Occupied(x) => {
+                f(unsafe { x.map.get_unchecked_mut(x.key.clone()) });
+                Entry::Occupied(x)
+            }
+            e @ Entry::Vacant(_) => e,
+        }
+    }
+}
+
+/// a
+#[derive(Debug)]
+pub struct OccupiedEntry<'a, K: Key, V> {
```

Comment:
You have a lot of `unsafe` `get_unchecked` stuff everywhere. Couldn't this be avoided by defining our `OccupiedEntry` in terms of `HashMap`'s `OccupiedEntry`?